src/java.base/share/classes/java/util/stream/GathererOp.java

```diff
@@ -0,0 +1,745 @@
+                    state = initializer.get();
+                collectorState = collectorSupplier.get();
+                proceed = true;
+            }
+
+            @ForceInline
+            Sequential evaluateUsing(Spliterator<T> spliterator) {
+                if (greedy)
+                    spliterator.forEachRemaining(this);
+                else
+                    do {
+                    } while (proceed && spliterator.tryAdvance(this));
+
+                return this;
+            }
+
+            /*
+             * No need to override isKnownDone() as the default is `false`
+             * and collectors can never short-circuit.
+             */
+            @Override
+            public boolean push(R r) {
+                collectorAccumulator.accept(collectorState, r);
+                return true;
+            }
+
+            @Override
+            public void accept(T t) {
+                /*
+                 * Benchmarking has shown that, in this case, conditional
+                 * writing of `proceed` is desirable  and if that was not the
+                 *  case, then the following line would've been clearer:
+                 *
+                 * proceed &= integrator.integrate(state, t, this);
+                 */
+
+                var ignore = integrator.integrate(state, t, this)
+                             || (!greedy && (proceed = false));
+            }
+
+            @SuppressWarnings("unchecked")
+            public CR get() {
+                final var finisher = gatherer.finisher();
+                if (finisher != Gatherer.<A, R>defaultFinisher())
+                    finisher.accept(state, this);
+                // IF collectorFinisher == null -> IDENTITY_FINISH
+                return (collectorFinisher == null)
+                           ? (CR) collectorState
+                           : collectorFinisher.apply(collectorState);
+            }
+        }
+
+        /*
+         * It could be considered to also go to sequential mode if the
+         * operation is non-greedy AND the combiner is Gatherer.defaultCombiner()
+         * as those operations will not benefit from upstream parallel
+         * preprocessing which is the main advantage of the Hybrid evaluation
+         * strategy.
+         */
+        if (!parallel)
+            return new Sequential().evaluateUsing(spliterator).get();
+
+        // Parallel section starts here:
+
+        final var combiner = gatherer.combiner();
+
+        /*
+         * The following implementation of hybrid parallel-sequential
+         * Gatherer processing borrows heavily from ForeachOrderedTask,
+         * and adds handling of short-circuiting.
+         */
+        @SuppressWarnings("serial")
+        final class Hybrid extends CountedCompleter<Sequential> {
+            private final long targetSize;
+            private final Hybrid leftPredecessor;
+            private final AtomicBoolean cancelled;
```

Comment:
It could be a volatile boolean and a comment is needed saying this field is only be accessed if greedy is true. And inside the constructors this should be the last field of the constructor to be initialized

Reply:
No, unfortunately it can't be a volatile boolean since it is a shared instance amongst a hierarchy.